pkg/agent/controller/bgp/controller_test.go

```diff
@@ -0,0 +1,1991 @@
+				65000,
+				"192.168.77.100",
+				[]string{podIPv6CIDR.String()},
+				[]bgp.PeerConfig{ipv6Peer1Config},
+			),
+			expectedCalls: func(mockBGPServer *bgptest.MockInterfaceMockRecorder) {
+				mockBGPServer.Start(ctx)
+				mockBGPServer.AddPeer(ctx, ipv6Peer1Config)
+				mockBGPServer.AdvertiseRoutes(ctx, []bgp.Route{{Prefix: podIPv6CIDR.String()}})
+			},
+		},
+		{
+			name:        "IPv4 & IPv6, as effective BGPPolicy, not advertise any Service IP due to no local Endpoint",
+			ipv4Enabled: true,
+			ipv6Enabled: true,
+			bpsToAdd: []runtime.Object{generateBGPPolicy(bgpPolicyName1,
+				creationTimestamp,
+				nodeLabels1,
+				1179,
+				65001,
+				true,
+				true,
+				true,
+				false,
+				false,
+				[]v1alpha1.BGPPeer{ipv4Peer1, ipv6Peer1})},
+			objects: []runtime.Object{
+				ipv4ClusterIP2,
+				ipv4ClusterIP2Eps,
+				ipv6ClusterIP2,
+				ipv6ClusterIP2Eps,
+				node,
+			},
+			expectedState: generateBGPPolicyState(1179,
+				65001,
+				nodeIPv4Addr.IP.String(),
+				nil,
+				[]bgp.PeerConfig{ipv4Peer1Config, ipv6Peer1Config},
+			),
+			expectedCalls: func(mockBGPServer *bgptest.MockInterfaceMockRecorder) {
+				mockBGPServer.Start(ctx)
+				mockBGPServer.AddPeer(ctx, ipv4Peer1Config)
+				mockBGPServer.AddPeer(ctx, ipv6Peer1Config)
+			},
+		},
+		{
+			name:        "IPv4, as alternative BGPPolicy",
+			ipv4Enabled: true,
+			bpsToAdd: []runtime.Object{generateBGPPolicy(bgpPolicyName2,
+				creationTimestamp, // As the effective BGPPolicy because the creationTimestamp is the oldest.
+				nodeLabels1,
+				179,
+				65000,
+				true,
+				false,
+				false,
+				false,
+				false,
+				[]v1alpha1.BGPPeer{ipv4Peer1}),
+				generateBGPPolicy(bgpPolicyName1,
+					creationTimestampAdd1s,
+					nodeLabels1,
+					179,
+					65000,
+					true,
+					false,
+					false,
+					false,
+					false,
+					[]v1alpha1.BGPPeer{ipv4Peer1})},
+			objects: []runtime.Object{ipv4ClusterIP1, ipv4ClusterIP1Eps, node},
+			existingState: generateBGPPolicyState(179,
+				65000,
+				nodeIPv4Addr.IP.String(),
+				[]string{ipStrToPrefix(clusterIPv4)},
+				[]bgp.PeerConfig{ipv4Peer1Config},
+			),
+			expectedState: generateBGPPolicyState(179,
+				65000,
+				nodeIPv4Addr.IP.String(),
+				[]string{ipStrToPrefix(clusterIPv4)},
+				[]bgp.PeerConfig{ipv4Peer1Config},
+			),
+		},
+	}
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newFakeController(t, tt.objects, append(tt.crdObjects, tt.bpsToAdd...), tt.ipv4Enabled, tt.ipv6Enabled)
+
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+			c.startInformers(stopCh)
+
+			// Fake the BGPPolicy state and the passwords of BGP peers.
+			c.bgpPolicyState = tt.existingState
+			if c.bgpPolicyState != nil {
+				c.bgpPolicyState.bgpServer = c.mockBGPServer
+			}
+			c.bgpPeerPasswords = bgpPeerPasswords
+
+			// Wait for the dummy event triggered by BGPPolicy add events, and mark it done.
+			waitAndDoneDummyEvent(t, c)
```

Comment:
I don't know if it matters for your tests, but I believe that to simulate what the controller does, it would be better to let the test case call `c.queue.Done` appropriately (after calling `syncBGPPolicy`), rather than calling it immediately in `waitAndDoneDummyEvent`.

Reply:
I have corrected all of them. It makes sense to call `c.queue.Done` after calling `syncBGPPolicy`, although it can also work to call `c.queue.Done` before calling `syncBGPPolicy`.